template-coq/src/pluginCore.ml

```diff
@@ -1,57 +1,167 @@
+    let evm, (c, _) = Constrintern.interp_casted_constr_evars_impls env evm hole (EConstr.of_constr ty) in
+    Obligations.check_evars env evm;
+    let obls, _, c, cty = Obligations.eterm_obligations env nm evm 0 (EConstr.to_constr evm c) ty in
+    (* let evm = Evd.minimize_universes evm in *)
+    let ctx = Evd.evar_universe_context evm in
+    let hook = Lemmas.mk_hook (fun _ gr _ ->
+        let env = Global.env () in
+        let evm = Evd.from_env env in
+        let evm, t = Evd.fresh_global env evm gr in
+        match Constr.kind t with
+        | Constr.Const (tm, _) ->
+          success env evm (Names.Constant.canonical tm)
+        | _ -> failwith "Evd.fresh_global did not return a Const") in
+    ignore (Obligations.add_definition nm ~term:c cty ctx ~kind ~hook obls) }
 
 let tmFreshName (nm : ident) : ident tm =
-  not_implemented "tmFreshName"
+  { run_tm = fun env evd success fail ->
+    let name' =
+      Namegen.next_ident_away_from nm (fun id -> Nametab.exists_cci (Lib.make_path id))
+    in success env evd name' }
 
 let tmAbout (nm : ident) : global_reference option tm =
-  not_implemented "tmAbout"
-let tmCurrentModPath (_ : unit) : string tm =
-  not_implemented "tmCurrentModPath"
+  { run_tm = fun env evd success fail ->
+    try
+      let qualid = Libnames.qualid_of_ident nm in
+      let gr = Smartlocate.locate_global_with_alias (CAst.make qualid) in
+      success env evd (Some gr)
+    with
+    | Not_found -> success env evd None }
+
+let tmCurrentModPath (_ : unit) : Names.ModPath.t tm =
+  { run_tm = fun env evd success _ ->
+    let mp = Lib.current_mp () in success env evd mp }
+
+let tmQuoteInductive (kn : kername) : mutual_inductive_body option tm =
+  { run_tm = fun env evm success fail ->
+        try
+          let mind = Environ.lookup_mind (Names.MutInd.make1 kn) env in
+          success env evm (Some mind)
+        with
+          _ -> success env evm None }
 
-let tmQuoteInductive (kn : kername) : _ tm =
-  not_implemented "tmQuoteInductive"
 let tmQuoteUniverses : _ tm =
   not_implemented "tmQuoteUniverses"
-let tmQuoteConstant (kn : kername) (bypass : bool) : _ tm =
-  not_implemented "tmQuoteConstant"
 
-let tmMkInductive : _ -> _ tm =
-  not_implemented "tmMkInductive"
+(* get the definition associated to a kername *)
+let tmQuoteConstant (kn : kername) (bypass : bool) : constant_entry tm =
+  { run_tm = fun env evd success fail ->
+        let cnst = Environ.lookup_constant (Names.Constant.make1 kn) env in
+        success env evd cnst }
+
+let tmInductive (mi : mutual_inductive_entry) : unit tm =
+  not_implemented "tmInductive"
 
 let tmExistingInstance (kn : kername) : unit tm =
-  not_implemented "tmExistingInstance"
-let tmInferInstance (tm : term) : term option tm =
-  not_implemented "tmInferInstance"
+  { run_tm = fun env evd success fail ->
+        (* note(gmm): this seems wrong. *)
+        let ident = Names.Id.of_string (Names.KerName.to_string kn) in
```

Comment:
This seems wrong. What is the right way to do this? Should `tmExistingInstance` take a `kername` or something else?

Reply:
So with #110, it is clear that you will get a `qualid` here, then you have to finish resolving it. In the non extracted monad it was done by:
```
Classes.existing_instance true (CAst.make (Libnames.Qualid (Libnames.qualid_of_ident (unquote_ident name)))) None
```
Can't you do the same?

Reply:
I can do the same, it seems like you saying that there is not a `Classes` API that works on fully resolved names, so knowing a `kername` is not very helpful in this context.

Reply:
Indeed, Coq’s API is not the best in that case. You can do resolution on the client side and transform to Qualid, when we fix the API it should take a global_reference instead

Reply:
@mattam82 is there a general rule for when something should take a `global_reference` and when something should take a `kername`? Assuming that you can't forge a `kername` the two types seem (semantically) isomorphic. Is that true or is there something that I am missing?

Reply:
`global_reference` is the user-friendly view of kernel names, they (shouldn't) be forged either.

Reply:
So yes, they play (almost) the same role. The thing is that global references give you the kind of definition you are talking about, while KerNames.t (in Coq's kernel) can also be used as global identifier for other things (e.g. tactics).

Reply:
Ah, I didn't realize that they could reference tactics as well. I thought the kernel didn't know anything about tactics. One, slightly odd thing, is that `global_reference` doesn't carry the name does it? Is it easy to get a `kername` from a `global_reference`?

Reply:
The kernel doesn't, it is actually the library manager that handles creation of the unique identifiers. I think in most cases yes, `global_reference` is to be preferred.

Reply:
The odd thing about `global_reference` is that it is more than a reference, it actually contains the value that it refers to, at least in the case of `IndRef` and `ConstructRef`.

Reply:
I’m not sure what you mean, inductive and constructors are referred to by a ˋkername/Names.MutInd.t` + some int indexes, so they are resolved names.

Reply:
Oh, I see. I don't have the types  in context.